File: UniTrade/app/src/main/java/com/example/unitrade/PurchaseManager.java
package com.example.unitrade;

import android.util.Log;
import com.google.firebase.Timestamp;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.WriteBatch;
import java.util.HashMap;
import java.util.Map;

public class PurchaseManager {
    private static final String TAG = "PurchaseManager";

    public interface PurchaseCallback {
        void onSuccess(String purchaseId);
        void onFailure(String error);
    }

    /**
     * Process a single purchase and automatically mark product as sold
     */
    public static void processPurchase(
            String productId,
            String buyerId,
            String sellerId,
            double price,
            String receivingMethod,
            String deliveryAddress,
            PurchaseCallback callback
    ) {
        FirebaseFirestore db = FirebaseFirestore.getInstance();

        // Create a batch write for atomic operation
        WriteBatch batch = db.batch();

        // Step 1: Create purchase document
        String purchaseId = db.collection("purchases").document().getId();
        Purchase purchase = new Purchase(productId, buyerId, sellerId, price);
        purchase.setId(purchaseId);
        purchase.setReceivingMethod(receivingMethod);
        purchase.setDeliveryAddress(deliveryAddress);

        batch.set(db.collection("purchases").document(purchaseId), purchase);

        // Step 2: Update product status
        Map<String, Object> productUpdates = new HashMap<>();
        productUpdates.put("status", price > 0 ? "sold" : "donated");
        productUpdates.put("soldAt", Timestamp.now());
        productUpdates.put("soldTo", buyerId);
        productUpdates.put("purchaseId", purchaseId);
        productUpdates.put("transactionDate", System.currentTimeMillis());

        batch.update(db.collection("products").document(productId), productUpdates);

        // Commit the batch
        batch.commit()
                .addOnSuccessListener(aVoid -> {
                    Log.d(TAG, "Purchase processed successfully: " + purchaseId);
                    callback.onSuccess(purchaseId);
                })
                .addOnFailureListener(e -> {
                    Log.e(TAG, "Failed to process purchase", e);
                    callback.onFailure(e.getMessage());
                });
    }

    /**
     * Process multiple purchases (for cart checkout)
     */
    public static void processMultiplePurchases(
            java.util.List<String> productIds,
            String buyerId,
            String receivingMethod,
            String deliveryAddress,
            PurchaseCallback callback
    ) {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        WriteBatch batch = db.batch();

        final int[] completed = {0};
        final int[] failed = {0};
        final StringBuilder errorMessages = new StringBuilder();

        // Process each product
        for (String productId : productIds) {
            // First get the product to know the seller and price
            db.collection("products")
                    .document(productId)
                    .get()
                    .addOnSuccessListener(doc -> {
                        Product product = doc.toObject(Product.class);
                        if (product == null) {
                            failed[0]++;
                            errorMessages.append("Product ").append(productId).append(" not found. ");
                            checkCompletion(completed, failed, productIds.size(), errorMessages, callback);
                            return;
                        }

                        // Create purchase
                        processPurchase(
                                productId,
                                buyerId,
                                product.getSellerId(),
                                product.getPrice(),
                                receivingMethod,
                                deliveryAddress,
                                new PurchaseCallback() {
                                    @Override
                                    public void onSuccess(String purchaseId) {
                                        completed[0]++;
                                        Log.d(TAG, "Product " + productId + " purchased successfully");
                                        checkCompletion(completed, failed, productIds.size(), errorMessages, callback);
                                    }

                                    @Override
                                    public void onFailure(String error) {
                                        failed[0]++;
                                        errorMessages.append(error).append(" ");
                                        checkCompletion(completed, failed, productIds.size(), errorMessages, callback);
                                    }
                                }
                        );
                    })
                    .addOnFailureListener(e -> {
                        failed[0]++;
                        errorMessages.append("Failed to load product ").append(productId).append(". ");
                        checkCompletion(completed, failed, productIds.size(), errorMessages, callback);
                    });
        }
    }

    private static void checkCompletion(
            int[] completed,
            int[] failed,
            int total,
            StringBuilder errorMessages,
            PurchaseCallback callback
    ) {
        if (completed[0] + failed[0] == total) {
            if (failed[0] == 0) {
                callback.onSuccess("All purchases completed");
            } else if (completed[0] > 0) {
                callback.onFailure("Some purchases failed: " + errorMessages.toString());
            } else {
                callback.onFailure("All purchases failed: " + errorMessages.toString());
            }
        }
    }
}

File: app/src/main/java/com/example/unitrade/PurchaseManager.kt
package com.example.unitrade

import android.util.Log
import com.example.unitrade.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*

object PurchaseManager {
    private const val TAG = "PurchaseManager"

    interface PurchaseCallback {
        fun onSuccess(purchaseId: String)
        fun onFailure(error: String)
    }

    fun processPurchase(
        productId: String,
        buyerId: String,
        sellerId: String,
        price: Double,
        receivingMethod: String,
        deliveryAddress: String,
        callback: PurchaseCallback
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Insert purchase
                val purchase = Purchase(productId, buyerId, sellerId, price).apply {
                    id = UUID.randomUUID().toString()
                    this.receivingMethod = receivingMethod
                    this.deliveryAddress = deliveryAddress
                }
                SupabaseClient.client.postgrest["purchases"].insert(purchase)

                // Update product
                val status = if (price > 0) "sold" else "donated"
                SupabaseClient.client.postgrest["products"].update({
                    set("status", status)
                    set("sold_at", Date())
                    set("sold_to", buyerId)
                    set("purchase_id", purchase.id)
                    set("transaction_date", System.currentTimeMillis())
                }) {
                    eq("id", productId)
                }

                withContext(Dispatchers.Main) {
                    Log.d(TAG, "Purchase processed successfully: ${purchase.id}")
                    callback.onSuccess(purchase.id!!)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    Log.e(TAG, "Failed to process purchase", e)
                    callback.onFailure(e.message ?: "Unknown error")
                }
            }
        }
    }

    fun processMultiplePurchases(
        productIds: List<String>,
        buyerId: String,
        receivingMethod: String,
        deliveryAddress: String,
        callback: PurchaseCallback
    ) {
        var completed = 0
        var failed = 0
        val errorMessages = StringBuilder()

        for (productId in productIds) {
            // Get product first
            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val products = SupabaseClient.client.postgrest["products"]
                        .select {
                            eq("id", productId)
                        }
                        .decodeList<Product>()
                    if (products.isNotEmpty()) {
                        val product = products[0]
                        processPurchase(
                            productId,
                            buyerId,
                            product.sellerId!!,
                            product.price,
                            receivingMethod,
                            deliveryAddress,
                            object : PurchaseCallback {
                                override fun onSuccess(purchaseId: String) {
                                    completed++
                                    Log.d(TAG, "Product $productId purchased successfully")
                                    checkCompletion(completed, failed, productIds.size, errorMessages, callback)
                                }

                                override fun onFailure(error: String) {
                                    failed++
                                    errorMessages.append(error).append(" ")
                                    checkCompletion(completed, failed, productIds.size, errorMessages, callback)
                                }
                            }
                        )
                    } else {
                        failed++
                        errorMessages.append("Product $productId not found. ")
                        checkCompletion(completed, failed, productIds.size, errorMessages, callback)
                    }
                } catch (e: Exception) {
                    failed++
                    errorMessages.append("Failed to load product $productId. ")
                    checkCompletion(completed, failed, productIds.size, errorMessages, callback)
                }
            }
        }
    }

    private fun checkCompletion(
        completed: Int,
        failed: Int,
        total: Int,
        errorMessages: StringBuilder,
        callback: PurchaseCallback
    ) {
        if (completed + failed == total) {
            if (failed == 0) {
                callback.onSuccess("All purchases completed")
            } else if (completed > 0) {
                callback.onFailure("Some purchases failed: $errorMessages")
            } else {
                callback.onFailure("All purchases failed: $errorMessages")
            }
        }
    }
}